src/pass/zero_elimination.cc

```diff
@@ -0,0 +1,1719 @@
+  if (combiner->result.size() != 1) {
+    return false;
+  }
+
+  if (!is_const_value(SuperSimplify(combiner->identity_element[0]), 0)) {
+    return false;
+  }
+
+  return is_const_value(SuperSimplify(combiner->result[0] -
+                                      (combiner->lhs[0] + combiner->rhs[0])),
+                        0);
+}
+
+// Return true if zero may be factored out of a reduction with this combiner.
+bool CanFactorZeroFromCombiner(const CommReducer& combiner, int value_index) {
+  if (!is_const_value(combiner->identity_element[value_index], 0)) {
+    return false;
+  }
+
+  Expr zero = make_zero(combiner->result[value_index].type());
+  Expr in = Substitute(combiner->result[value_index],
+                       {{combiner->lhs[value_index], zero},
+                        {combiner->rhs[value_index], zero}});
+  in = SuperSimplify(in);
+
+  return is_const_value(in, 0);
+}
+
+Expr InlineThisCall(const Expr& expr) {
+  if (const Call* op = expr.as<Call>()) {
+    if (op->call_type == Call::CallType::Halide) {
+      if (const ComputeOpNode* op_comp = op->func.as<ComputeOpNode>()) {
+        Array<Var> tensor_axes;
+        for (const auto& var : op_comp->axis) {
+          tensor_axes.push_back(var->var);
+        }
+
+        Stmt inlined = Inline(Evaluate::make(expr), op->func, tensor_axes,
+                              op_comp->body[op->value_index]);
+        if (const ir::Evaluate* ev = inlined.as<ir::Evaluate>()) {
+          // If it is a reduction, clone it
+          return CloneReduction(ev->value);
+        }
+      }
+    }
+  }
+
+  return expr;
+}
+
+Tensor InlineTailCall(const Tensor& tensor) {
+  return op::TransformBody(tensor, InlineThisCall);
+}
+
+class InlineTensorsMutator : public IRMutator {
+ public:
+  explicit InlineTensorsMutator(const Array<Tensor>& inlineable, bool inline_reductions = false)
+      : inline_reductions_(inline_reductions) {
+    for (const Tensor& tensor : inlineable) {
+      inlineable_.emplace(tensor->op.operator->(), tensor->value_index);
+    }
+  }
+
+  Expr Mutate_(const Call* op, const Expr& e) {
+    if (op->call_type == Call::CallType::Halide) {
+      const ComputeOpNode* op_comp = op->func.as<ComputeOpNode>();
+      if (inlineable_.empty() || inlineable_.count({op_comp, op->value_index})) {
+        if (op_comp && (inline_reductions_ || !op_comp->body[0].as<Reduce>())) {
+          Array<Var> tensor_axes;
+          for (const auto& var : op_comp->axis) {
+            tensor_axes.push_back(var->var);
+          }
+
+          Stmt inlined = Inline(Evaluate::make(e), op->func, tensor_axes,
+                                op_comp->body[op->value_index]);
+          if (const ir::Evaluate* ev = inlined.as<ir::Evaluate>()) {
+            // If it is a reduction, clone it
+            return Mutate(ev->value);
+          }
+        }
+      }
+    }
+
+    return e;
+  }
+
+ private:
+  std::set<std::pair<const OperationNode*, int>> inlineable_;
```

Comment:
Unless we look at the code, it's hard to say that `int` refers to. A short comment would help.